M-map full chunks of Head from disk

codesome opened the pull request:
tl-dr desc for the PR from @krasi-georgiev 

-----------------
When appending to the head and a chunk is full it is flushed to the disk and m-mapped (memory mapped) to free up memory

Prom startup now happens in these stages
 - Iterate the m-maped chunks from disk and keep a map of series reference to its slice of mmapped chunks.
- Iterate the WAL as usual. Whenever we create a new series, look for it's mmapped chunks in the map created before and add it to that series.


If a head chunk is corrupted the currpted one and all chunks after that are deleted and the data after the corruption is recovered from the existing WAL which means that a corruption in m-mapped files results in NO data loss.


[Mmaped chunks format](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/head_chunks.md)  - main difference is that the chunk for mmaping now also includes series reference because there is no index for mapping series to chunks.
[The block chunks](https://github.com/prometheus/prometheus/blob/master/tsdb/docs/format/chunks.md) are accessed from the index which includes the offsets for the chunks in the chunks file - example - chunks of series ID have offsets 200, 500 etc in the chunk files.
In case of mmaped chunks, the offsets are stored in memory and accessed from that. During WAL replay, these offsets are restored by iterating all m-mapped chunks as stated above by matching the series id present in the chunk header and offset of that chunk in that file.

**Prombench results**

_WAL Replay_

1h Wal reply time
30% less wal reply time - 4m31 vs 3m36
2h Wal reply time
20% less wal reply time - 8m16 vs 7m

_Memory During WAL Replay_

[High Churn](http://prombench.prometheus.io/grafana/d/7gmLoNDmz/prombench?orgId=1&var-RuleGroup=All&var-pr-number=6679&from=1588177559275&to=1588182978872)	
10-15% less RAM -  32gb vs 28gb
20% less RAM after compaction 34gb vs 27gb
[No Churn](http://prombench.prometheus.io/grafana/d/7gmLoNDmz/prombench?orgId=1&var-RuleGroup=All&var-pr-number=6679&from=1588199791787&to=1588201716740)
20-30% less RAM -  23gb vs 18gb
40% less RAM after compaction 32.5gb vs 20gb

Screenshots are in [this comment](https://github.com/prometheus/prometheus/pull/6679#issuecomment-621678932)

-----------------------

Prerequisite: https://github.com/prometheus/prometheus/pull/6830 (Merged)

Closes https://github.com/prometheus/prometheus/issues/6377. More info in the linked issue _and_ the doc in that issue _and_ the doc inside that doc inside that issue :)


- [x] Add tests
- [x] Explore possible ways to get rid of new globals added in `head.go`
- [x] Wait for https://github.com/prometheus/prometheus/pull/6830 to be merged
- [x] Fix windows tests

codesome commented:
Forgot to open as a draft PR, changed the title to WIP.

codesome commented:
I have removed the WIP tag now and it is ready for review. The windows tests are failing and I could not find a way yet to get rid of the failure where the files are being removed in `defer` in the unit tests.

I will run prombench for a few hours tomorrow on this to see the benefit and also stress test this PR.

codesome commented:
/prombench master

bwplotka commented on tsdb/chunks/chunks.go:
```diff
@@ -190,40 +190,50 @@ func (w *Writer) cut() error {
+
+	w.files = append(w.files, f)
+	if w.wbuf != nil {
+		w.wbuf.Reset(f)
```
No need for `else` here - just return is enough to simplify.

codesome replied:
It is required :) only one of them should be run

bwplotka commented on tsdb/chunks/chunks.go:
```diff
@@ -190,40 +190,50 @@ func (w *Writer) cut() error {
+	return nil
+}
+
+func cutSegmentFile(dirFile *os.File, chunksFormat byte, preallocate bool, segmentSize int64) (headerSize int, newFile *os.File, seq int, err error) {
```
I would avoid boolean arguments to functions. It's not extensible and can be confusing. I think in this case you can just not preallocate if `segmentSize` is 0, right?

bwplotka commented on tsdb/chunks/chunks.go:
```diff
@@ -190,40 +190,50 @@ func (w *Writer) cut() error {
 	if err != nil {
 		return err
 	}
+	w.n = int64(n)
```
Why we do this outside of cutSegment File?

codesome replied:
`cutSegmentFile` is shared by this struct and also the new `HeadReadWriter` that I introduced. It is a standalone function now and not a part of either of the struct. Hence need to do it outside.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+
+const (
+	// DefaultHeadChunkSegmentTime is the default chunks segment time range.
+	DefaultHeadChunkSegmentTime = 15 * time.Minute
```
:+1: 

I would add a comment why 15 minutes.

codesome replied:
Thinking more about it, maybe 15m isn't right. Assuming an average scrape interval of 15s, the chunks would be completed after 30m, so with 15m as the default, there would be many empty files. I will make it 40m with the reasoning in the comments.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+		return nil, err
+	}
+
+	// mmap existing files
```
Not a full sentence

codesome replied:
A cleanup which I missed, that comment isn't relevant now

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+	return hrw, nil
+}
+
+// NewDirReader returns a new Reader against sequentially numbered files in the
```
Comment is wrong

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+		w.bsMtx.Unlock()
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxSegmentSize))
```
Something is wrong with this `w.bsMtx`.

Imagine that here at this point `w.Close` is happening from another goroutine and only after that below steps are executed:

```
	w.bsMtx.Lock()
	w.cs[w.curFileSequence] = mmapFile
	w.bs[w.curFileSequence] = realByteSlice(mmapFile.Bytes())
	w.bsMtx.Unlock()
```
..  new opened mmap file to be not closed, leaking the file descriptor.

I think we should lock the whole `cut` method.

codesome replied:
Locking the entire method would create problems in the queries which access these chunks. I added another way to take care of this: lock the `wbufLock` lock used in `WriteChunk` (which calls `cut`) inside `Close` too, and additionally check for `Close` being called inside `WriteChunk` and exit early. Does this looks fine now?

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+		closers = append(closers, w.cs[seq])
+		delete(w.bs, seq)
+		delete(w.cs, seq)
+		if err := os.Remove(segmentFile(w.dirFile.Name(), seq)); err != nil {
```
What about the `w.size`? Should we substract the size of those?

codesome replied:
Good point. We should not only subtract here but also add when we cut a new segment file :) Thanks!

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+		chkStart = int((ref << 32) >> 32)
+	)
+
+	// TODO: Fix this indexing
```
What do you mean by fix here?

codesome replied:
Leftover comment :) irrelevant now.

bwplotka commented on tsdb/head.go:
```diff
@@ -350,7 +368,67 @@ func (h *Head) updateMinMaxTime(mint, maxt int64) {
+	)
+	go func() {
+		defer close(decoded)
+		for r.Next() {
```
What about `r.Error()`?

codesome replied:
Thanks! Missed it

bwplotka commented on tsdb/head.go:
```diff
@@ -571,31 +620,106 @@ func (h *Head) Init(minValidTime int64) error {
+		return errors.Wrap(err, "iterate on-disk chunks")
+	}
 
+	level.Info(h.logger).Log("msg", "replaying WAL, this may take awhile")
```
hm, we do lot's of code even before that logline. Should we move it to the beginning of this method?

Also can we add `finished replaying WAL` with the extra field `elapsed` that measures elapsed time? It's always useful for long-running task (:

bwplotka commented on tsdb/head.go:
```diff
@@ -1673,25 +1818,34 @@ func (s *memSeries) maxTime() int64 {
+	if s.headChunk != nil {
+		chunkRef, err := chunkReadWriter.WriteChunk(s.ref, s.headChunk.minTime, s.headChunk.maxTime, s.headChunk.chunk)
+		if err != nil {
+			panic(err)
```
panic? Can we do better?

codesome replied:
I forgot to put a TODO here. We should be able to get rid of the panic, I will look into it.

codesome replied:
There is a panic below too in the same function for the appender. And looking through the flow where this is called, hard fail seems to be the pattern in the ingest path if it fails. I will keep it as panic for now.

brian-brazil commented on tsdb/chunks/chunks.go:
```diff
@@ -190,40 +190,50 @@ func (w *Writer) cut() error {
 	}
-	if err = w.dirFile.Sync(); err != nil {
-		return err
+	if err = dirFile.Sync(); err != nil {
```
We don't currently have an fsync on any critical path, this could stall things for many many seconds.

codesome replied:
What would you suggest doing here?

brian-brazil commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+	return merr.Err()
+}
+
+// IterateAllChunks iterates on all the chunks in it's byte slices in the order of the segment file sequence
```
its

brian-brazil commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+// IterateAllChunks iterates on all the chunks in it's byte slices in the order of the segment file sequence
+// and runs the provided function on each chunk. It returns on the first error encountered.
+func (w *HeadReadWriter) IterateAllChunks(f func(seriesRef, chunkRef uint64, mint, maxt int64) error) error {
+	// Iterate files in ascending order.
```
Do we check for gaps?

codesome replied:
We do not. The gap could be for many reasons and I think we should not check for gaps.

brian-brazil replied:
What sort of reasons? Gaps aren't possible for the normal WAL, and we check for gaps. I don't see why we wouldn't be doing the same reason here.

codesome replied:
I think I misunderstood what a gap is here, so, what do you mean by gap here?

brian-brazil replied:
If the files are say 1, 3 and 4. 2 is missing.

codesome replied:
Ahh, yes. I definitely misunderstood before. I will make it error out in that case. Once https://github.com/prometheus/prometheus/pull/6129 is in, I can follow up with a similar logic.

brian-brazil commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+}
+
+// WriteChunk writes chunk in the following format:
+// | Series Ref <8B> | MinT <8B> | MaxT <8B> | Chunk Encoding <1B> | Chunk Data Length <varint> | Chunk Data |
```
I'd prefer we keep this in the docs, rather than duplicating it in the comments and one of them going stale.

brian-brazil commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+		return err
+	}
+	defer func() {
+		if err != nil {
```
Which value of `err` exactly?

I think you want a typical `defer f.Close()` here.

codesome replied:
We don't want to close the file if there is no error. I will change the error names to make it more clear, I am confused too here regarding which `err` it is :)

brian-brazil replied:
Add a comment then too, as that's not the usual pattern.

brian-brazil commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+	oldSeq := w.curFileSequence
+	w.curFileSequence = seq
+
+	// Write current time in milliseconds for mint and 0 for maxt.
```
Why are we writing a walltime inside the tsdb?

codesome replied:
Whoops, had not thought much when I first implemented, it's definitely wrong. I will combine the suggestion of taking maxt from samples and write mint from samples too by changing it to `cut(t)`.

brian-brazil commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+	if w.wbuf != nil {
+		w.wbuf.Reset(f)
+	} else {
+		w.wbuf = bufio.NewWriterSize(f, 1024)
```
This seems like rather a small buffer

codesome replied:
Buffer is flushed for every chunk write, won't 1024B be enough for 1 chunk?

brian-brazil replied:
Writing chunk by chunk doesn't sound the most efficient, I'd expect us to buffer up a fair bit more.

codesome replied:
I considered that, but the reading part gets more complicated and also we have to give a chunk reference from `WriteChunk`. Additionally, if there is a crash while chunks are still in the buffer, we would lose a lot of data with time range depending on the scrape interval.

codesome replied:
The additional complexity could be taken care of, but the data loss during any crash is my main concern.

brian-brazil replied:
Wouldn't the data be in the WAL?

codesome replied:
_Right!_

codesome replied:
I have added some in-memory buffer support now. Some race needs to be taken care, but I feel this will cause a lot of lock contention between reads and writes.

brian-brazil commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,511 @@
+		return err
+	}
+
+	// Writing maxt of the file in milliseconds.
```
I'd expect maxt to be of the data, not walltime.

codesome replied:
Yes, will change it.

brian-brazil commented on tsdb/head.go:
```diff
@@ -571,31 +620,106 @@ func (h *Head) Init(minValidTime int64) error {
-			return errors.Wrap(err, "open checkpoint")
+	// Backfill series from the checkpoint first if it exists.
+	if err := h.executeOnCheckpointReader(func(checkpointReader *wal.Reader) error {
+		// A corrupted checkpoint is a hard error for now and requires user
```
We should treat it like any other wal corruption, delete what is corrupted onwards and continue to startup

codesome replied:
This is the same old comment what was present, haven't changed anything here in terms of processing the checkpoint. Did you mean for the chunk files that we should delete anything after the corruption?

brian-brazil replied:
Yes

codesome replied:
So the corruption in the chunks files is not obvious. There are no records to decode specifically that we can error on. The error is more like `"not enough bytes to read the next chunk"`. In that case, we are not even sure if all the chunks read till that part is correct, so we have to discard all the chunks from that file, which sounds tricky after having parsed all the (right/wrong) chunks till then. So a safe fallback that I see is to discard all the chunks that we read till now and recover entirely from WAL.

brian-brazil commented on tsdb/head.go:
```diff
@@ -1211,6 +1341,9 @@ func (h *Head) compactable() bool {
 
 // Close flushes the WAL and closes the head.
 func (h *Head) Close() error {
+	if err := h.chunkReadWriter.Close(); err != nil {
```
I suspect you need to close the wal even if this fails.

codesome replied:
Yes

codesome commented:
@geekodour helped me run another benchmark on this with a little less churn than default prombench, you can find the links to the dashboard in this pr https://github.com/prometheus-community/prometheus/pull/50. This will give an idea of improvement for a saner Prometheus setup (do note that prombench uses 5s as scrape interval so the memory difference is visible early on in both prombench instances than what would be on a typical setup if I assume 15s scrape interval)

codesome commented:
**Benchmark results are here**

1. With a high churn (which was run on this PR), with series ranging between 4.3M to 9.8M because of churn, the RSS savings was ranging between 10-22%.
![more_churn](https://user-images.githubusercontent.com/15064823/73329689-96b85300-4284-11ea-9cb1-ae58cc4fa6f1.png)

2. With a slightly lower churn (which was run on [another PR](https://github.com/prometheus-community/prometheus/pull/50) but with the same code), with series ranging between 2.2M to 3.2M because of churn, the RSS savings was ranging between 30-45%.
![less_churn](https://user-images.githubusercontent.com/15064823/73329701-9ddf6100-4284-11ea-85ed-24f89c2d2d60.png)

**Some conclusions**

1. Memory reduction is as expected. With higher churn, as there are more incomplete chunks, the gain shows slightly lower as incomplete chunks stay in the head (incomplete=didn't touch 120 samples). With a lower churn, the gain was higher. On a Prometheus setup with very little churn, the gain would be even higher. 

**Some observations (graphs are from high churn benchmark):**

1. The number of Head chunks look weird for this PR, it slightly grew compared to master at a constant rate. When it came down suddenly, there was a spike in the number of active series for a small while before it became normal (it shot up to 27.7M then came back to the normal 5.6M). This needs to be investigated.
![Screenshot from 2020-01-29 10-59-29](https://user-images.githubusercontent.com/15064823/73330372-99b44300-4286-11ea-9460-37d0bfaebbd7.png)

codesome commented:
/prombench cancel

brian-brazil commented:
One thought: How will this interact with isolation? Should we do isolation first?

codesome commented:
> One thought: How will this interact with isolation? Should we do isolation first?

M-map of chunks can be treated as a blackbox for that I believe as at the end the active chunk is still in the memory and when accessing the chunk you get the same old chunk.

codesome commented:
For the reviewers: I have changed the title again to WIP. The things I will be doing in this WIP is:

1. I have implemented the repair. Hence add more tests around it.
2. Debug the increase in chunks that we saw in https://github.com/prometheus/prometheus/pull/6679#issuecomment-579604539
3. Breakdown this PR into some small parts which are safe to do so (like the `noopWAL` in tests, some refactoring, etc).

I will let you know when it is ready for review again.

codesome commented:
I think I found the subtle bug (3rd pic in https://github.com/prometheus/prometheus/pull/6679#issuecomment-579604539) that was causing the dashboard to show the wrong number of removed chunks (and potentially also wrong chunk id accesses). I have fixed it in this commit https://github.com/prometheus/prometheus/pull/6679/commits/e4444b508a79f4c752d58768ba355f79e7b59f87 and will run benchmark again once I have removed the WIP.

codesome commented:
I think I am done with the changes that I wanted to do. I am running the benchmark again for some-time on this high churn benchmark to see if the chunk count is fixed.

/prombench master

bwplotka commented:
Awesome! Will look tomorrow :+1:

codesome commented:
Chunk count looks fine now. Memory gains stand between 10-13% in this test, mostly due to churn I believe which is expected. I have removed the WIP tag, ready for review again.

Suggested review flow as this is a big PR:
1. `head_chunks.go` which has the main logic of this PR, and it's unit tests.
2. Then `head.go` for the changes to interact with mmap chunks.
    * Reviewing `head.go` first by taking the mmapping logic as blackbox would also be good I guess.
3. And then all the other files which are mainly because of the function signature changes.

/prombench cancel

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,708 @@
+	inBufferChunks      [inBufferShards]map[uint64]chunkenc.Chunk
+	inBufferChunksLocks [inBufferShards]sync.RWMutex
+
+	// Reader
```
Missing Trailing period.

bwplotka commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,708 @@
+	return err
+}
+
+func (w *HeadReadWriter) cut(mint int64) (returnErr error) {
```
why not just `err`?

codesome replied:
It was not clear which error was being used in the defer, hence, just to be clear

bwplotka commented on tsdb/chunks/head_chunks_test.go:
```diff
@@ -0,0 +1,282 @@
+
+func TestHeadReadWriter_WriteChunk(t *testing.T) {
+	hrw, close := testHeadReadWriter(t)
+	defer close()
```
just put in one defer please (same for others)

bwplotka commented on tsdb/docs/format/head_chunks.md:
````diff
@@ -0,0 +1,41 @@
+
+```
+┌──────────────────────────────┐
+│  magic(0x85BD40DD) <4 byte>  │
````
Just double-checking, I hope it is a different magic than our other files, right?

codesome replied:
I am re-using the same code to writing the initial headers of the segment files. So it will be the same.

bwplotka replied:
I think this is wrong. Magic code is the ID of the format. Since it's different than chunk segments, it has to have unique magic code IMO (:

codesome replied:
There is a version too for the file, so doesn't the version decide the format?

bwplotka replied:
No. Think of magic code as "Type" of format."Version" as instance of it.

https://en.wikipedia.org/wiki/Magic_number_(programming)

> A constant numerical or text value used to identify a file format or protocol; for files, see List of file signatures

codesome replied:
Okay. I will add a different magic number then. (This will be done in the new PR opened)

bwplotka commented on tsdb/docs/format/head_chunks.md:
```diff
@@ -0,0 +1,41 @@
+├──────────────────────────────┤
+│    padding(0) <3 byte>       │
+├──────────────────────────────┤
+│    mint       <8 byte>       │
```
But what type? int? uint? float? Same for  above

codesome replied:
Will add.

codesome replied:
Will add.

bwplotka commented on tsdb/docs/format/head_chunks.md:
````diff
@@ -0,0 +1,41 @@
+
+```
+┌─────────────────────┬───────────────┬───────────────┬───────────────────┬───────────────┬──────────────┐
+| series ref <8 byte> | mint <8 byte> | maxt <8 byte> | encoding <1 byte> | len <uvarint> | data <bytes> |
````
ditto

codesome commented:
> put new head writer and head_test in separate PR without hooking it to TSDB head yet.. WDYT?

Definitely, if that helps

codesome commented:
I have addressed the reviews, but I will be splitting the PR as you mentioned now.

codesome commented:
Opened the PR here https://github.com/prometheus/prometheus/pull/6830

bwplotka commented:
Thank You!

krasi-georgiev commented:
how is the wal reply looking with this change?

codesome commented:
@krasi-georgiev Good point. I haven't tested that yet but it is going to add a tiny bit of replay time. Will do a benchmark for that.

krasi-georgiev commented:
haven't looked at the PR, but the proposal doc suggests that this will drastically reduce the wal reply, isn't this true with this implementation?

codesome commented:
> but the proposal doc suggests that this will drastically reduce the wal reply, isn't this true with this implementation?

This implementation, nope. Currently during replay we do that following:
1. Iterate over the entire checkpoint + WAL once to recover the series in the memory. Iteration step takes few seconds if we just look at the record type.
2. Recover the mmapped chunks.
3. At last, do the usual WAL replay while skipping the series record as we already did that in step 1.

[This is the reason](https://github.com/prometheus/prometheus/issues/6377#issuecomment-575170954) why we iterate over the entire WAL twice.

Hence with this the replay time will slightly increase but not decrease. I plan to do some follow up testing with writing partial chunks on disk and speeding up replay with that.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -0,0 +1,708 @@
+
+	for i, b := range w.bs {
+		if b.Len() < HeadSegmentHeaderSize {
+			return errors.Wrapf(errInvalidSize, "invalid segment header in segment %d", i)
```
Note to self: have repair for these errors too after https://github.com/prometheus/prometheus/pull/6830 is merged

roidelapluie commented:
@beorn7 @brian-brazil do you think this could balance the overhead of isolation? Should we target it for 2.17.0 ?

beorn7 commented:
I'm not a TSDB expert. But this here seems to be an optimization for a Prometheus server under memory pressure. It's not really working in the direction where the isolation creates overhead.

roidelapluie commented:
> I'm not a TSDB expert. But this here seems to be an optimization for a Prometheus server under memory pressure. It's not really working in the direction where the isolation creates overhead.

But with isolation it will become under memory pressure :)

beorn7 commented:
There is a bit of memory increase in Prombench, but the production server I looked at doesn't really have different memory behavior.

But even if it does, mmap'ing things like here doesn't really remove the need for the memory, it just makes it easier for the OS to swap out currently unneeded stuff to disk. I wouldn't really sell the one as a compensation for the other.

roidelapluie commented:
> There is a bit of memory increase in Prombench, but the production server I looked at doesn't really have different memory behavior.
> 
> But even if it does, mmap'ing things like here doesn't really remove the need for the memory, it just makes it easier for the OS to swap out currently unneeded stuff to disk. I wouldn't really sell the one as a compensation for the other.

Yeah so it will prevent some OOM by slowing down stuff.

codesome commented:
Finally, tests pass in windows :tada:

codesome commented:
And it's ready for review! For the reviewers:

1. For the many methods added around the WAL replay, it is recommended to look at `Head.Init` first to see how they are used, and then jump into their implementation later.
2. Adding new mmap chunk is in `memSeries.cut` where we create new chunk.
3. The rest of the part is mostly to accommodate the removal of chunks slice.

(If you want to check the mmaping implementation, it is in https://github.com/prometheus/prometheus/pull/6830)

codesome commented:
The test failure seems to be unrelated, I see it happening in other PRs too.

codesome commented on web/web_test.go:
```diff
@@ -90,7 +90,6 @@ func TestGlobalURL(t *testing.T) {
 }
 
 func TestReadyAndHealthy(t *testing.T) {
-	t.Parallel()
```
Last time I checked, the tests were conflicting with each other because of the disk use. Let me check if that is still the case.

codesome replied:
So there is a global variable in head.go which is causing a race. We could get rid of that global, but we would need to store a pointer with every memSeries. That would be around 7.6KiB per million series. I guess the tradeoff looks fine (tradeoff with having a global variable).

codesome commented on tsdb/head.go:
```diff
@@ -258,7 +271,7 @@ func (h *Head) PostingsCardinalityStats(statsByLabelName string) *index.Postings
 // A larger stripeSize will allocate more memory up-front, but will increase performance when handling a large number of series.
 // A smaller stripeSize reduces the memory allocated, but can decrease performance with large number of series.
-func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, chunkRange int64, stripeSize int) (*Head, error) {
+func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, chunkRange int64, chkDirRoot string, pool chunkenc.Pool, stripeSize int) (*Head, error) {
```
As an arg so that WAL need not be initialised. WAL can be nil in the tests.

And about chunk pool, yes, the pool is actually shared between block querier and compactor currently. The sad part is, nothing is being put in the pool yet, so there is no re-use it. I made a small attempt to make it re-use the chunks, but it is not trivial, so had halted my attempt.

krasi-georgiev replied:
hm, yeah I see what you mean. I am now thinking should the chunks even be in the wal directory. Maybe it would be much better that these are in the root directory.

codesome replied:
Sound fine to keep it in the root directory as we always need mmapping and cannot be disabled if WAL is disabled.

codesome commented on tsdb/db_test.go:
```diff
@@ -870,9 +882,17 @@ func TestWALSegmentSizeOptions(t *testing.T) {
-			if _, err := os.Stat(filepath.Join(dbDir, "wal")); !os.IsNotExist(err) {
-				t.Fatal("wal directory is present when the wal is disabled")
-			}
+			// Check that WAL dir is there.
```
WAL cannot be disabled by a user I believe (except the Read only DB). It is only in tests that WAL is disabled at places. And based on how code would be designed for simplicity, it would be easier to always have mmaped chunk rather than it being optional.

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -512,7 +512,7 @@ func (cdm *ChunkDiskMapper) Chunk(ref uint64) (chunkenc.Chunk, error) {
 // and runs the provided function on each chunk. It returns on the first error encountered.
```
Because we want to do some operations on the chunks while iterating :) for example in our case replay the chunks in Head.

codesome replied:
https://github.com/prometheus/prometheus/blob/fd4521d2ebac6e467dd35db1e38c0e97aad49921/tsdb/head.go#L713-L742

codesome commented on tsdb/chunks/head_chunks.go:
```diff
@@ -577,7 +577,10 @@ func (cdm *ChunkDiskMapper) IterateAllChunks(f func(seriesRef, chunkRef uint64,
-			idx += n + int(dataLen) // Skip the data.
+			idx += n
+
+			numSamples := binary.BigEndian.Uint16(mmapFile.byteSlice.Range(idx, idx+2))
```
That is how our XOR chunks is designed as I see it https://github.com/prometheus/prometheus/blob/fd4521d2ebac6e467dd35db1e38c0e97aad49921/tsdb/chunkenc/xor.go#L77-L80

krasi-georgiev replied:
yeah, now I see it. we should really extend the doc to explain the format of the data field inside the chunks as it would be useful. Anyway it is not related to this PR, but thanks for the pointer.

codesome commented on tsdb/head.go:
```diff
@@ -78,29 +82,33 @@ type Head struct {
-	cardinalityCache      *index.PostingsStats // Posting stats cache which will expire after 30sec.
-	lastPostingsStatsCall time.Duration        // Last posting stats call (PostingsCardinalityStats()) time for caching.
+	cardinalityCache      *index.PostingsStats // posting stats cache which will expire after 30sec
+	lastPostingsStatsCall time.Duration        // last posting stats call (PostingsCardinalityStats()) time for caching
```
Yes

codesome commented on tsdb/head_test.go:
```diff
@@ -1635,3 +1753,19 @@ func testHeadSeriesChunkRace(t *testing.T) {
+	h, err := NewHead(nil, nil, wlog, chunkRange, wlog.Dir(), nil, DefaultStripeSize)
+	testutil.Ok(t, err)
+
+	testutil.Ok(t, h.chunkDiskMapper.IterateAllChunks(func(_, _ uint64, _, _ int64, _ uint16) error { return nil }))
```
It is called in `head.Init()`. Thing is, not all the tests that use this Head call the Init() method. So, in order to set the mint and maxt of mmap chunk files for those tests, we are calling this here with no action on chunks.

krasi-georgiev replied:
so why not call `head.Init()`  here?

codesome replied:
Different tests have different use cases with Init (like different Mint etc), so it would not be right to call init here.

codesome commented on tsdb/head_test.go:
```diff
@@ -42,10 +41,11 @@ import (
+	h, _, closer := getTestHead(b, 10000, false)
+	defer closer()
+	defer func() {
+		testutil.Ok(b, h.Close())
```
Some tests need to close the head at different places, so can't make it generic to close it in `closer`

krasi-georgiev commented:
> Ok, I reviewed the easy parts, but before going to the actual implementation, I am wondering why not use mapped chunks for everything? I mean why do we still need the memChunk? The mmaper has its own buffer so can adjust it to write to disk at very low intervals to reduce the write amplification.

@codesome ^

brian-brazil commented:
Is that safe in the face of partial writes by the kernel?

codesome commented:
Oops missed it. So, we need the memChunk fir the active chunk. I don't think mmaping an active chunk is simple enough or even efficient. And what Brian mentioned.

krasi-georgiev commented:
>  I don't think mmaping an active chunk is simple enough or even efficient. And what Brian mentioned.

I didn't mean to write an active chunk but to put it in the mapped type chunk buffer.

The mapped type chunks are written only when the buffer is exhausted. With a big buffer are you just reading and writing to the memory and no disk is involved.

codesome commented:
The right buffer size cannot be decided beforehand. Imagine a case when there is a surge of number of series, in that case, buffer will fill up with chunks and the older chunks which are still active will be evicted and flushed from the buffer.

krasi-georgiev commented:
alright, yes I see what you mean. Yes I now see the reason to keep those as separate structs.

codesome commented on tsdb/head.go:
```diff
@@ -1796,27 +2043,30 @@ func (s *memSeries) chunkID(pos int) int {
+	}
+	if len(s.mmappedChunks) > 0 {
+		for i, c := range s.mmappedChunks {
+			if c.maxTime >= mint {
```
This slice works like the `chunks` slice just that they are mmapped. So the chunk gets appended to the back of the slice whenever a new chunk is cut, and that happens in the time sorted order. And mmapping happens in the same order, hence the replay of mmaping chunks are also in time sorted order. (Additionally, we check for out of order chunks when replaying mmapped chunks)

codesome commented on tsdb/head.go:
```diff
@@ -572,32 +626,176 @@ func (h *Head) Init(minValidTime int64) error {
+		h.metrics.mmapChunkCorruptionTotal.Inc()
+		// Repair is best effort here. If it fails, data will be recovered from WAL.
+		// Hence we wont lose any data (given WAL is not corrupt).
+		// TODO(codesome): add test for this repair.
```
Yes, will remove that comment

codesome commented on tsdb/head.go:
```diff
@@ -572,32 +626,176 @@ func (h *Head) Init(minValidTime int64) error {
-		return errors.Wrap(err, "find last checkpoint")
+	level.Info(h.logger).Log("msg", "creating series from WAL")
+	if err := h.loadSeriesFromWAL(refSeries, multiRef); err != nil {
+		// TODO(codesome): If corruption is found here and repaired, we still need to load the WAL.
```
This is a TODO for this PR itself, we cannot go ahead without this repair. I will add it.

codesome commented on tsdb/head.go:
```diff
@@ -572,32 +626,176 @@ func (h *Head) Init(minValidTime int64) error {
+	return nil
+}
+
+func (h *Head) loadMmappedChunks(refSeries map[uint64]*memSeries, multiRef map[uint64]uint64) error {
```
Right! Not needed. And I will add a comment. I need these 2 variables to be part of Head now to take care of WAL replay after corruption in series loading phase. I will add comments there.

codesome commented on tsdb/head.go:
```diff
@@ -1825,7 +2075,11 @@ func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkC
+		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t {
+			return false, chunkCreated
+		}
+		c = s.cut(t, chunkDiskMapper)
```
I would say splitting is going to make it more complex to understand. The way you can look at it is:
1. `cut()` is called to create a new chunk. We are checking for `nil` here because if this was the first sample being appended, then we want to create a new chunk before we proceed.
2. The non-nil chunk inside `cut()` is because we don't want to mmap the in-memory chunk. If we are going to create a new chunk and if that is going to be in-memory chunk, we don't mmap it. If there was already a chunk (`s.head() != nil`), then we m-map it.

Comparing this nil check with the nil check inside `cut()` are not much related as their purpose is different. I will add comments.

krasi-georgiev replied:
What about something like this?


```
func (s *memSeries) cut(mint int64, chunkDiskMapper *chunks.ChunkDiskMapper) *memChunk {
	if s.headChunk != nil {
		// There is already a chunk, m-map it before we replace it with a new one.
		chunkRef, err := chunkDiskMapper.WriteChunk(s.ref, s.headChunk.minTime, s.headChunk.maxTime, s.headChunk.chunk)
		if err != nil {
			if err == chunks.ErrChunkDiskMapperClosed {
				// The head is being closed, hence chunkDiskMapper is closed.
				// Return the same head chunk as there wont be more samples appended anyway.
				return s.headChunk
			}
			panic(err)
		}
		s.mmappedChunks = append(s.mmappedChunks, &mmappedChunk{
			ref:        chunkRef,
			numSamples: uint16(s.headChunk.chunk.NumSamples()),
			minTime:    s.headChunk.minTime,
			maxTime:    s.headChunk.maxTime,
		})
	}

	return s.newHead(mint)
}

func (s *memSeries) newHead(mint int64) *memChunk {
	s.headChunk = &memChunk{
		chunk:   chunkenc.NewXORChunk(),
		minTime: mint,
		maxTime: math.MinInt64,
	}

	// Set upper bound on when the next chunk must be started. An earlier timestamp
	// may be chosen dynamically at a later point.
	s.nextAt = rangeForTimestamp(mint, s.chunkRange)

	app, err := s.headChunk.chunk.Appender()
	if err != nil {
		panic(err)
	}
	s.app = app
	return s.headChunk
}
```

then this part of the code will simple become as 
```
if c == nil {
		// Out of order sample. Don't cut a chunk with wrong timestamp.
		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t {
			return false, chunkCreated
		}
		// There is no chunk in this series yet, create the first chunk.
		c = s.newHead(t)
		chunkCreated = true
	}
```

codesome replied:
I still very much prefer to have a single method to create a new chunk, makes it much simpler to call it anywhere and no need to think if we have a head or no before calling `cut`. I am still not convinced that the check for head chunk being nil is a problem here. Like some methods have some small tasks to perform at the beginning on a method, similarly here we just mmap a chunk when head chunk already exists. I will move the mmaping part to a different method and call it inside cut if that makes it simpler; to have a single way to create new chunk.

krasi-georgiev replied:
yep that works as well, I could follow the code much better, thanks

krasi-georgiev commented on tsdb/head.go:
```diff
@@ -1298,34 +1556,42 @@ func (h *headChunkReader) Chunk(ref uint64) (chunkenc.Chunk, error) {
+	c, garbageCollect := s.chunk(int(cid), h.head.chunkDiskMapper)
+	defer func() {
+		if garbageCollect {
+			c.chunk = nil
```
hm , how does this work, if it is garbage collected here how would the caller read the data?

codesome replied:
I think the comments are confusing. Here we are garbage collecting the `c`, which is `*memChunk`, while the reader uses `c.chunk`. So setting c.chunk `nil` does not garbage collect `c.chunk` as it is referenced somewhere else already. Note that we are collecting `c` after `c.chunk` is already used somewhere.

codesome replied:
Does that make sense?

krasi-georgiev replied:
hm , no sorry I didn't understand what you mean.

codesome replied:
Here `c` and `c.chunk` are 2 different things (basically 2 pointers). After calling `s.chunk()`, we are mainly using `c.chunk`, so we should not garbage collect `c.chunk`. But the struct that is holding it, `c`, can be garbage collected (added back to the pool). Garbage collecting it does not affect `c.chunk`, and setting `c.chunk = nil` also does not affect `c.chunk` as that means changing the pointer and not the value inside the struct pointed by the pointer.

krasi-georgiev replied:
I get it now, this explains it very simple
https://play.golang.org/p/BX_zJ5_QaYx

krasi-georgiev commented on tsdb/head.go:
```diff
@@ -572,32 +636,216 @@ func (h *Head) Init(minValidTime int64) error {
+		return err
+	}
+
+	// No corruption in loading series. Hence set these to nil to mark that.
```
Would you mind if we use a dedicated variable with a more descriptive name for this.
or can try to use the metric to detect/indicate if there was a repair.

codesome replied:
I will use a dedicated variable to indicate this, we need this info in `PostWALRepairRecovery`

krasi-georgiev replied:
so if you have a dedicated variable why do you still need to set `refSeries` and `multiRef` to nil?

also what do you mean by 

`Hence set these to nil to garbage collect it.`

codesome replied:
We need `refSeries` for perf reasons: not to load series by holding all the locks again.
We need to hold `multiRef` because that is created only when we replay series and we need it to replay samples properly. I have added an explanation in the comments here https://github.com/prometheus/prometheus/blob/d5e7aed61e4eedbd7daf48a5f204433051ea0efc/tsdb/head.go#L94-L96

> Hence set these to nil to garbage collect it.

Go will garbage collect the memory for the items which are not referenced anywhere in code currently (IIUC), so if we don't set it to nil, it will always be held in memory. Setting it to `nil` allows Go GC to collect it. I will mention Go GC in comment there, as we have other GC as well.

krasi-georgiev replied:
Yes I was confused that you are talking about the internal garbage collection in the code

Why reason why this wouldn't work in the `PostWALRepairRecovery`>

```
if !h.seriesReplayCorruption {
	return nil
}
``

codesome replied:
It will, I will change it

codesome commented:
> I can't seem to see where is the logic to NOT load samples from the wal at startup that are already present in the mapped disk chunks

We skip the out of order samples in the append which effectively takes care of what you said above

https://github.com/prometheus/prometheus/blob/f7dc9ed5e5f0e5b5c89e0cba9c8a489b85d2f7c4/tsdb/head.go#L2131-L2134
https://github.com/prometheus/prometheus/blob/f7dc9ed5e5f0e5b5c89e0cba9c8a489b85d2f7c4/tsdb/head.go#L2141-L2144

codesome commented on tsdb/head.go:
```diff
@@ -1825,7 +2075,11 @@ func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkC
 
 	if c == nil {
-		c = s.cut(t)
+		// Out of order sample. Don't cut a chunk with wrong timestamp.
```
Should it be a metric?

krasi-georgiev replied:
I would say a log message like  the unknownRefs

codesome replied:
I see that append returns `false` only when it is out of order. Considering that, I think it would be easy to add.

codesome commented on tsdb/head.go:
```diff
@@ -1877,8 +2131,15 @@ func computeChunkEndTime(start, cur, max int64) int64 {
+		}
+	}()
+
 	// TODO(fabxc): Work around! A querier may have retrieved a pointer to a
```
Might be, I am not sure how the `head.gc()` call and queries are linked. I would leave this investigation outside to some other time. But with m-mapping, I think this must be solved.

codesome commented on tsdb/head.go:
```diff
@@ -1896,12 +2157,18 @@ func (s *memSeries) iterator(id int, isoState *isolationState, it chunkenc.Itera
-		for j, d := range s.chunks {
-			totalSamples += d.chunk.NumSamples()
+		for j, d := range s.mmappedChunks {
+			totalSamples += int(d.numSamples)
```
Lock is held when getting the iterator
https://github.com/prometheus/prometheus/blob/f7dc9ed5e5f0e5b5c89e0cba9c8a489b85d2f7c4/tsdb/head.go#L1593-L1596

Similarly when appending
https://github.com/prometheus/prometheus/blob/f7dc9ed5e5f0e5b5c89e0cba9c8a489b85d2f7c4/tsdb/head.go#L1289-L1293

krasi-georgiev replied:
and this call is during the wal reply so can't the iterator can't  be called during the wal reply

https://github.com/prometheus/prometheus/pull/6679/files#diff-6fb13507bfadf9819fea5bda61f599e6R344

maybe add a comment that `append` and `iterator` are nos safe to call concurently.

codesome replied:
Yes, I skipped mentioning replay because no queries happen at that time. I will add the comments.

codesome commented on tsdb/head.go:
```diff
@@ -1896,12 +2157,18 @@ func (s *memSeries) iterator(id int, isoState *isolationState, it chunkenc.Itera
-			totalSamples += d.chunk.NumSamples()
+		for j, d := range s.mmappedChunks {
+			totalSamples += int(d.numSamples)
 			if j < ix {
```
For such involved internal details, I don't think so.

codesome commented on tsdb/head.go:
```diff
@@ -1896,12 +2157,18 @@ func (s *memSeries) iterator(id int, isoState *isolationState, it chunkenc.Itera
 		}
+		// mmappedChunks does not contain the last chunk. Hence check it separately.
+		if len(s.mmappedChunks) < ix {
+			previousSamples += s.headChunk.chunk.NumSamples()
```
I added some additional test cases in existing isolation tests, but mostly for the restart and replay of mmap chunks. But looking at the existing test cases, I think it is covered. @beorn7 should confirm (basically testing isolation when samples extend beyond the last chunk).

beorn7 replied:
It looks like it @codesome , but this whole PR is way too involved to allow a real confirmation from my side. I invested 15m to see if I can wrap my head around it or if I can just look at the new isolation test cases and come to a conclusion based on that. I'm afraid I would need hours to even get close to a sufficient understanding. I don't have that time right now. sorry…

codesome replied:
No issues @beorn7. I am confident enough that it doesn't cause issues with the isolation, so I guess it should be fine.

krasi-georgiev commented:
> > I can't seem to see where is the logic to NOT load samples from the wal at startup that are already present in the mapped disk chunks
> 
> We skip the out of order samples in the append which effectively takes care of what you said above
> 
> https://github.com/prometheus/prometheus/blob/f7dc9ed5e5f0e5b5c89e0cba9c8a489b85d2f7c4/tsdb/head.go#L2131-L2134
> 
> 
> https://github.com/prometheus/prometheus/blob/f7dc9ed5e5f0e5b5c89e0cba9c8a489b85d2f7c4/tsdb/head.go#L2141-L2144


What do you mean by "Don't cut a chunk with wrong timestamp."?

Shouldn't this comment be just like:
`// Don't append samples that are already in the mmapped chunks.`

on a different note for such cases, lets return `return false, false`. Main reason is that if it is `return false, chunkCreated` than I have to look in the whole function to see if it is changed anywhere.

codesome commented:
`wrong` is a wrong word here, I will update the comment. There are 2 different things here, one is appending samples and another is actually cutting a chunk which does not require appending sample. I will make it clear in the comment (and also the returns)

krasi-georgiev commented:
> `wrong` is a wrong word here, I will update the comment. There are 2 different things here, one is appending samples and another is actually cutting a chunk which does not require appending sample. I will make it clear in the comment (and also the returns)

and why the comment that I suggested is not accurate?
maybe something like:

`// Sample is already in the mmaped chunk files to don't append it or cut a new head chunk.`
The main reason is so that it becomes more obvious that the samples that are already in the mmaped chunks are not loaded from the wal.

Then the comment: 
`// There is no chunk in this series yet, create the first chunk.`

becomes obsolete.

codesome commented:
I made it `// Out of order sample. Don't cut a chunk for a sample that we wont append.`. We cannot says it's in the mmapped just based on the time overlap as it might be some other sample whose timestamp is within the mmaped chunks.

krasi-georgiev commented:
I pinged you on IRC for a follow up question.

krasi-georgiev commented on tsdb/head.go:
```diff
@@ -572,32 +645,217 @@ func (h *Head) Init(minValidTime int64) error {
+// PostWALRepairRecovery is used to recover the remaining samples from WAL if the corruption
+// occured in the series loading phase. This must be called iff Head.Init encountered a WAL corruption
+// and a repair was called on the WAL.
+// TODO(codesome): Add a test for this.
```
is this still a todo?

codesome replied:
Yes, I will add it eventually before this PR is merged (after we have reviewed the verified the other integrations)

krasi-georgiev commented:
/prombench master

codesome commented:
/prombench stop

Small thing to fix

codesome commented:
/prombench cancel

codesome commented:
/prombench master

krasi-georgiev commented:
/prombench cancel

roidelapluie commented on tsdb/head.go:
```diff
@@ -80,27 +83,44 @@ type Head struct {
+	chunksRemoved            prometheus.Counter
+	gcDuration               prometheus.Summary
+	samplesAppended          prometheus.Counter
+	outOfOrderSamples        prometheus.Counter
```
This is not registered.

krasi-georgiev replied:
hm, wait isn't there a test for this metric somewhere in the tests? how did we miss that?

codesome replied:
There is no test for this. I can write one if it is necessary.

roidelapluie replied:
thanos has tests to see if all metrics are registered

codesome replied:
I will add a test. I had missed incrementing the metric when we return `storage.ErrOutOfBounds`

codesome replied:
I added a test now, and I have split the metric into 2, 1 for out of order (within compaction window but out of order), 1 for out of bound (beyond the compaction window)

codesome replied:
I had this in mind but thought I will follow up with another PR after this to keep the diff size smaller. I would still prefer to do it in a separate PR.

bwplotka replied:
Cool, up to you

codesome commented:
Some good results from Prombench summarised by @krasi-georgiev 

**WAL Replay**

_1h Wal reply time_
* 30% less wal reply time - 4m31 vs 3m36

_2h Wal reply time_
* 20% less wal reply time - 8m16 vs 7m

**Memory usage**

_With high churn_

* 10-15% less RAM initially
* 15-20% less RAM after compaction

![Screenshot from 2020-04-30 13-23-29 (copy)](https://user-images.githubusercontent.com/15064823/80686281-1fde6080-8ae6-11ea-9705-08dff41cf0fe.png)

_With 0 churn_

* 20-30% less RAM initially
* 35-40% less RAM after compaction

![Screenshot from 2020-04-30 13-23-32 (copy)](https://user-images.githubusercontent.com/15064823/80686288-2371e780-8ae6-11ea-9307-d12b0f7aa0fa.png)

codesome commented:
@brian-brazil suggested a change to avoid iterating over the WAL twice during replay, and I am not sure why did that not hit me before :upside_down_face: 

Current replay:
1. Iterate over the WAL and only process the series records, hence loading all the empty series in memory.
2. Replay back the mmapped chunks and attach them to the relevant series.
3. Iterate the WAL again and replay the remaining samples.

New suggested change:
1. Replay the mmapped chunks and hold the map of series ID to the mmapped chunks in the memory.
2. Do the normal WAL replay as we have now, and whenever we come across a series record, attach the mmapped chunks from the map to that series.

I will make this change on Monday (or weekend) and run prombench again to test the replay. This is going to remove a lot of additional complexity brought by splitting up the WAL replay!

codesome commented on tsdb/head.go:
```diff
@@ -472,7 +516,11 @@ func (h *Head) loadWAL(r *wal.Reader, multiRef map[uint64]uint64) (err error) {
 
-				if !created {
+				if created {
+					if mc, ok := mmappedChunks[series.ref]; ok {
+						series.mmappedChunks = mc
+					}
```
I have a question here: Consider the series was not created and it was an old one, should the mmapped chunks be taken for `series.ref` (already present series) or from `s.Ref` (the reference from new series record), as the reference of the series active before shutting down would be from the last series record.

codesome replied:
Or more like it is possible we need to club the mmapped chunks for both `series.ref` and `s.Ref`

codesome replied:
So I have kept it as a TODO for the future as there are some races to take care when we are removing the old data when we come across a duplicate record. As for now, the WAL replay will take care of adding back all the data and Head gc will take care of removing old data at the end.

codesome commented:
/prombench master

To test the replay times after the last change

codesome commented:
/prombench cancel

krasi-georgiev commented on tsdb/head.go:
```diff
@@ -1776,12 +1940,32 @@ func (s *memSeries) appendable(t int64, v float64) error {
 
-func (s *memSeries) chunk(id int) *memChunk {
+// chunk returns the chunk for the id. If garbageCollect is true, it means that the returned *memChunk
+// (and not the chunkend.Chunk inside it) can be garbage collected after it's usage.
```
```suggestion
// (and not the chunk.Chunk inside it) can be garbage collected after it's usage.
```

codesome replied:
Should be `chunkenc`, will fix

codesome commented on tsdb/head.go:
```diff
@@ -629,6 +688,55 @@ func (h *Head) Init(minValidTime int64) error {
+
+// repairMmappedChunks attempts repair of the mmapped chunks and if it fails, it clears all the previously
+// loaded mmapped chunks.
+func (h *Head) repairMmappedChunks(err error) map[uint64][]*mmappedChunk {
```
If we want to be consistent, then `wal.Repair` should also be renamed, because WAL repair basically deletes the corrupted data. Sounds good? 

Also, I kept is separate to keep away the inner details from Init and make Init more readable on a higher level, so I would prefer keeping it separate.

codesome replied:
I renamed only for mmapped chunks now, WAL can be another PR, lots of places to change for that

krasi-georgiev commented:
lgtm

bwplotka commented on tsdb/head.go:
```diff
@@ -1776,12 +1940,32 @@ func (s *memSeries) appendable(t int64, v float64) error {
+		if err == chunks.ErrChunkDiskMapperClosed {
+			return nil, false
+		}
+		panic(err)
```
do  we really want panic here?

codesome replied:
I dont see other clean ways to do it, returning error would cause a lot more places to be changed and something to be thought/designed over. Additionally, if there is an error, then there is something definitely wrong (data corruption), so a hard fail looks fine here.

bwplotka replied:
Not really sure, do we really want panic whole ingestion and query because of single map chunk file failed read?

codesome replied:
So the only blocker here is the [`Chunk` interface where we have `Iterator(Iterator) Iterator`](https://github.com/prometheus/prometheus/blob/532f7bbac92acedc1994b14130a3737cd88a7b6d/tsdb/chunkenc/chunk.go#L43-L68). We have this call here where we call that `.chunk` method https://github.com/prometheus/prometheus/blob/532f7bbac92acedc1994b14130a3737cd88a7b6d/tsdb/head.go#L1893-L1894
So in order to be able to propagate this error here, we need to change the interface and a whole lot of places. I will take a look after fixing the windows test fail.

codesome replied:
Turns out changing the interface is a big task :) do you have any suggestions here which would not involve changing the interface?

bwplotka replied:
metric, log and ignore? :thinking:

bwplotka replied:
Are we sure it's not recoverable error?

bwplotka replied:
Let's add TODO at least for now

codesome replied:
> Are we sure it's not recoverable error?

I think so. The error occurs whenever there is [corruption in data](https://github.com/prometheus/prometheus/blob/532f7bbac92acedc1994b14130a3737cd88a7b6d/tsdb/chunks/head_chunks.go#L434-L509). So in case of corruption, metric, log and ignore might not be the right thing to do, which will also lead to gaps in queries and maybe increase the time to recover the faulty disk.

I will add a TODO.

bwplotka commented on tsdb/head.go:
```diff
@@ -1792,27 +1976,31 @@ func (s *memSeries) chunkID(pos int) int {
 		s.headChunk = nil
-	} else {
-		s.headChunk = s.chunks[len(s.chunks)-1]
+		s.mmappedChunks = nil
```
Looks like we can return here, right?

codesome replied:
Yes, but it makes the next `if` a noop, so I think it should be fine for a small method.

bwplotka replied:
Early return is really what is more cleaner IMO (:

codesome replied:
Sure, no issues with that

bwplotka commented on tsdb/head.go:
```diff
@@ -1821,7 +2009,12 @@ func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkC
 
 	if c == nil {
-		c = s.cut(t)
+		// Out of order sample. Sample timestamp is already in the mmaped chunks, so ignore it.
```
Let's put this comment inside this if - otherwise I was reading this as c== nil means out of order.

codesome replied:
Hah sorry for that! moving inside.

bwplotka commented on tsdb/head.go:
```diff
@@ -1837,7 +2030,7 @@ func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkC
 	}
 	if t >= s.nextAt {
-		c = s.cut(t)
+		c = s.cutNewHeadChunk(t, chunkDiskMapper)
```
Quite hard to deduce if we can have 2x cutNewHeadChunk (c== nil and t >= s.nextAt)

codesome replied:
Because we cut for `t` above, the `t` here is never going to be above `s.nextAt`, because it is set based on the timestamp passed when you cut a new chunk.

PS: it has been like that since forever, I have not changed anything, just the names.

bwplotka commented on tsdb/head.go:
```diff
@@ -1792,27 +1976,31 @@ func (s *memSeries) chunkID(pos int) int {
 // isolation for this append.)
-func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkCreated bool) {
+// It is unsafe to call this concurrently with s.iterator(...) without holding the series lock.
+func (s *memSeries) append(t int64, v float64, appendID uint64, chunkDiskMapper *chunks.ChunkDiskMapper) (success, chunkCreated bool) {
```
Ok, I need to admit this method is quite confusing for me. The return arguments are odd, and we really magically guess what is out of order error what is not. This will be hard to maintain and extend, so I would suggest small reflector. At least we should have consistent return arguments with proper `error` I think (so we can pass out of order error).

Even with comments I am having trouble to follow the flow and tell if there is any bug on edge cases etc, so something is wrong here ): Can we somehow simplify?

codesome replied:
I will add a comment saying when success is false (OOO), also, given `success` is false only for out of order, I will change the variable name in the return argument. As for returning errors, I would like to leave it as a follow-up work as it is not linked to this PR.

bwplotka commented on tsdb/head.go:
```diff
@@ -2035,3 +2245,14 @@ func (ss stringset) slice() []string {
+	minTime, maxTime int64
+}
+
+// Returns true if the chunk overlaps [mint, maxt].
```
```suggestion
// OverlapsClosedInterval returns true if the chunk overlaps [mint, maxt].
```

codesome commented on tsdb/db_test.go:
```diff
@@ -849,8 +849,14 @@ func TestWALSegmentSizeOptions(t *testing.T) {
+					files = append(files, f)
+				}
+			}
 			for _, f := range files[:len(files)-1] {
```
Will add a comment. We need all files except the last one to have the size of the option set, the last file should be less than the max for obvious reasons.

codesome commented on tsdb/db_test.go:
```diff
@@ -859,9 +865,15 @@ func TestWALSegmentSizeOptions(t *testing.T) {
+					files = append(files, f)
+				}
+			}
+			testutil.Assert(t, len(files) > 1, "current WALSegmentSize should result in more than a single WAL file.")
```
This is another test case, and if you see a little below, the last file is not ignored, it is just checked differently (to have the size less than the max segment size)

codesome commented on tsdb/head.go:
```diff
@@ -574,6 +631,17 @@ func (h *Head) Init(minValidTime int64) error {
+	mmappedChunks, err := h.loadMmappedChunks()
+	if err != nil {
+		level.Error(h.logger).Log("msg", "loading on-disk chunks failed", "err", err)
+		h.metrics.mmapChunkCorruptionTotal.Inc()
```
Right, I will add a check for the corruption error

codesome commented on tsdb/head.go:
```diff
@@ -629,6 +697,56 @@ func (h *Head) Init(minValidTime int64) error {
+		slice := mmappedChunks[seriesRef]
+		if len(slice) > 0 {
+			if slice[len(slice)-1].maxTime >= mint {
+				return errors.Errorf("out of sequence m-mapped chunk for series ref %d", seriesRef)
```
I actually used m-mapped/mmapped everywhere (maybe missed 1 or 2?). I am fine with changing it to more description, but I would like to keep it like that for now and follow up later (lots of places to check and change, which I dont want to get into right now)

codesome commented on tsdb/head.go:
```diff
@@ -1034,13 +1162,14 @@ func (a *headAppender) Commit() error {
 
 		if !ok {
 			total--
+			a.head.metrics.outOfOrderSamples.Inc()
```
Yes, series.append skips only when its out of order, I will add a comment.

codesome commented on tsdb/head.go:
```diff
@@ -1729,30 +1878,45 @@ func (s *memSeries) maxTime() int64 {
+
+func (s *memSeries) mmapCurrentHeadChunk(chunkDiskMapper *chunks.ChunkDiskMapper) {
+	if s.headChunk == nil {
+		// There is no head chunk, so nothing to m-map here.
```
From another comment: I am planning to keep it m-map* for now and follow up later for this change, there are many places to make this change. I have `memory-map*` for metrics for now as that is end user-facing.

codesome commented on tsdb/head.go:
```diff
@@ -1776,12 +1940,32 @@ func (s *memSeries) appendable(t int64, v float64) error {
 }
 
-func (s *memSeries) chunk(id int) *memChunk {
+// chunk returns the chunk for the id. If garbageCollect is true, it means that the returned *memChunk
```
chunk id, I will add that in the comment. And, I feel "from where" is not necessary, but will add anyway saying from in-memory or m-mapped.

bwplotka commented on tsdb/head.go:
```diff
@@ -1792,27 +1976,32 @@ func (s *memSeries) chunkID(pos int) int {
 // isolation for this append.)
-func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkCreated bool) {
+// It is unsafe to call this concurrently with s.iterator(...) without holding the series lock.
+// outOfOrder is false when the sample of out of order.
```
```suggestion

```
then?

bwplotka commented on tsdb/head.go:
```diff
@@ -1792,27 +1976,32 @@ func (s *memSeries) chunkID(pos int) int {
-func (s *memSeries) append(t int64, v float64, appendID uint64) (success, chunkCreated bool) {
+// It is unsafe to call this concurrently with s.iterator(...) without holding the series lock.
+// outOfOrder is false when the sample of out of order.
+func (s *memSeries) append(t int64, v float64, appendID uint64, chunkDiskMapper *chunks.ChunkDiskMapper) (ourOfOrder, chunkCreated bool) {
```
```suggestion
func (s *memSeries) append(t int64, v float64, appendID uint64, chunkDiskMapper *chunks.ChunkDiskMapper) (sampleInOrder, chunkCreated bool) {
```

codesome replied:
Whoops! Thanks

codesome commented:
Windows test is finally fixed in https://github.com/prometheus/prometheus/pull/6679/commits/eb2e6c181b175cedf0de3989758e0cb6e8d59091

bwplotka commented:
:tada:

brancz commented:
Incredible work! 🎉🎊
